src/array_like.rs

```diff
@@ -0,0 +1,111 @@
+//! TODO
+
+use crate::{
+    ndarray::{Array, Array1, ArrayView, Axis, Dimension},
+    Element, PyReadonlyArray,
+};
+use ndarray::{Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+use pyo3::{exceptions::PyValueError, FromPyObject, PyAny, PyResult};
+
+/// TODO
+#[derive(Debug)]
+pub struct PyArrayLike<'py, T, D>(ArrayLike<'py, T, D>)
+where
+    T: Element,
+    D: Dimension;
+
+#[derive(Debug)]
+enum ArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    Reference(PyReadonlyArray<'py, T, D>),
+    Converted(Array<T, D>),
+}
+
+impl<'py, T, D> PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    /// TODO
+    pub fn into_owned_array(self) -> Array<T, D> {
+        match self.0 {
+            ArrayLike::Converted(x) => x,
+            ArrayLike::Reference(x) => x.to_owned_array(),
+        }
+    }
+
+    /// TODO
+    pub fn as_view(&self) -> ArrayView<T, D> {
+        match &self.0 {
+            ArrayLike::Converted(x) => x.view(),
+            ArrayLike::Reference(x) => x.as_array(),
+        }
+    }
+
+    /// TODO
+    pub fn is_reference(&self) -> bool {
+        match &self.0 {
+            ArrayLike::Reference(_) => true,
+            ArrayLike::Converted(_) => false,
+        }
+    }
+}
+
+impl<'py, T, D> FromPyObject<'py> for PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+    Vec<T>: FromPyObject<'py>,
+{
+    fn extract(ob: &'py PyAny) -> PyResult<Self> {
+        if let Ok(array) = ob.extract::<PyReadonlyArray<T, D>>() {
+            return Ok(PyArrayLike(ArrayLike::Reference(array)));
+        }
+
+        if D::NDIM.unwrap_or(1) == 1 {
+            if let Ok(array) = ob.extract::<Vec<T>>() {
+                let res = Array1::from(array)
```

Comment:
I think using `Array1::from_vec` would be more explicit here and assure us that this is zero-copy.

That `impl<T> FromPyObject for Vec<T>` basically just iterates the Python value with a sequence type check added on top. So I wonder whether `Array1::from_iter` wouldn't be preferable (after successfully calling `PyAny::iter` on the Python value) because it is more generally and should be equally performant.

Reply:
Actually, `Array1::from_iter` just wraps `Array1::from_vec` and mapping the iterator makes it impossible to forward extraction errors.